.azure-pipelines/INSTALL.md

```diff
@@ -0,0 +1,112 @@
+During this installation step, warnings describing user access and legal comitments will be displayed like this:
+```
+!!! ACCESS REQUIRED !!!
+```
+
+This document suppose that the Azure DevOps organization is named _certbot_, and the Azure DevOps project is also _certbot_.
+
+## Useful links
+
+* https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=azure-devops&tabs=schema
+* https://www.azuredevopslabs.com/labs/azuredevops/github-integration/
+* https://docs.microsoft.com/en-us/azure/devops/pipelines/ecosystems/python?view=azure-devops
+
+## Prerequisites
+
+### Having a GitHub account
+
+Use your GitHub user for a normal GitHub account, or a user that has administrative rights to the GitHub organization if relevant.
+
+### Having an Azure DevOps account
+- Go to https://dev.azure.com/, click "Start free with GitHub"
+- Login to GitHub
+
+```
+!!! ACCESS REQUIRED !!!
+Personal user data (email + profile info, in read-only)
+```
+
+- Microsoft will create a Live account using the email referenced for the GitHub account. This account is also linked to GitHub account (meaning you can log it using GitHub authentication)
+- Proceed with account registration (birth date, country), add details about name and email contact
+
+```
+!!! ACCESS REQUIRED !!!
+Microsoft proposes to send commercial links to this mail
+Azure DevOps terms of service need to be accepted
+```
+
+_Logged to Azure DevOps, account is ready._
+
+### Installing Azure Pipelines to GitHub
+
+- On GitHub, go to Marketplace
+- Select Azure Pipeline, and "Set up a plan"
+- Select Free, then "Install it for free"
+- Click "Complete order and begin installation"
+
+```
+!!! ACCESS !!!
+Azure Pipeline needs RW on code, RO on metadata, RW on checks, commit statuses, deployements, issues, pull requests.
+Access can be defined for all or only selected repositories, which is nice.
+```
+
+- Redirected to Azure DevOps, select the account created in _Having an Azure DevOps account_ section.
+- Select the organization, and click "Create a new project" (let's name it the same than the targetted github repo)
+- The Visibility is public, to profit from 10 parallel jobs
+
+```
+!!! ACCESS !!!
+Azure Pipelines needs access to the GitHub account (in term of beeing able to check it is valid), and the Resources shared between the GitHub account and Azure Pipelines.
+```
+
+_Done. We can move to pipelines configuration._
+
+## Import an existing pipelines from `.azure-pipelines` folder
+
+- On Azure DevOps, go to your organization (eg. _certbot_) then your project (eg. _certbot_)
+- Click "Pipelines" tab
+- Click "New pipeline"
+- Where is your code? Select "GitHub"
+
+```
+!!! ACCESS !!!
+Here comes the big requests:
+- Admin access to webhooks and services (not the code)
+- RO access on personal data (email + profile)
+- RO + RW access to repositories (code, issue, PR, wiki, settings, webhooks, services, deploy keys, collaboration invites)
```

Comment:
Oh wow. So it basically needs full access to everything? I don't think there's much missing from this list.

Maybe we could give them this access, but given how much time we spent avoiding doing something similar for Docker Hub, I'm not immediately sure we want to do that.

Reply:
This is not the same situation than DockerHub.

First, we are talking here about a CI/CD tool that works on the GitHub project, not a tool working from the GitHub project like DockerHub. These access are quite relevant since its purpose is to proceed actions on the project. In fact, access for Travis CI and for Azure Pipelines are quite similar:
* Travis CI: `Read and write access to checks, commit statuses, deployments, and repository hooks`
* Azure Pipelines: `Read and write access to checks, commit statuses, deployments, issues, and pull requests`

Second, this is not a full access to everything. DockerHub was asking from a full access on the organization itself, but Azure Pipelines (or other CIs) asks for a write access on the repositories only. And this is only a user level access, not an admin one. Controls done on GitHub for protected branches, or merging policies, are out of its reach.

Third, and this is valid for all systems that uses the GitHub Applications: this is a satisfyingly secured approach to delegate authentication. We are not talking about long-live tokens giving full-access to the APIs, but short-live tokens generated dynamically during the authentication process. See https://developer.github.com/apps/migrating-oauth-apps-to-github-apps/

Fourth, versus the risk of migrating, we need to consider the advantages to move to Azure Pipelines. It is not only about gaining a lot of parallel executors, or efficient and stable VMs both for Windows and Linux. It is also to build a full CD solution with the appropriate solution. Modern CI/CD are pipelines: Travis and AppVeyor do not have this, while CircleCI, GitHub Actions and Azure Pipelines are designed around it. In the long term having dedicated pipelines is the best choice to do in my opinion.

Otherwise, the hacks like in `before_install`, `branches.only` or `not-on-master/extended-test-suite` we find in our Travis config will become more and more present as we go deeper in a CD approach (first with Windows installer, after certainly with Snaps or other deployed artifacts).

Reply:
> Azure Pipelines: Read and write access to checks, commit statuses, deployments, issues, and pull requests

This is not the line I'm concerned about or commented on.

> Second, this is not a full access to everything. DockerHub was asking from a full access on the organization itself, but Azure Pipelines (or other CIs) asks for a write access on the repositories only. And this is only a user level access, not an admin one. Controls done on GitHub for protected branches, or merging policies, are out of its reach.

I think there may have been a miscommunication here. My understanding of the section of this document I commented on talks about admin access for webhooks and services and write access on things like your GitHub settings.

Reply:
After discussions on 1-1, we agree on the fact that Travis vs Azure Pipelines on actual permissions, when compared from the GitHub settings looks very similar.

@bmw to install this PR on a fork and verify this similarity. We need to confirm in particular what are the `Read write access on GitHub settings` assertion that is encountered in popups to setup Azure Pipelines, since it does not seem to be related to anything in the GitHub settings.

If it is confirmed that Azure Pipelines does not get significant more permissions than Travis, it will unblock this PR merging.

Reply:
I just finished playing with this a bit.

When following the instructions here (or the most intuitive flow on Azure DevOps), Azure tries to set up an Oauth app in addition to the GitHub app we initially installed. This Oauth app is where the concerning permissions about write access to all aspects of your repository come in.

I tried to look into what these repository permissions allows Azure to do but couldn't find much. I think they correspond to the repository scope described [here](https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/) but I couldn't find anything about what exactly that entails.

(As a slightly unrelated aside, I looked into the permissions that Travis' Oauth app requires and it is:
```
Organizations and teams
Read-only access
This application will be able to read your organization, team membership, and private project boards.

Commit statuses
Read and write access
This application will be able to read and write commit statuses (no direct code access).

Deployments
Manage deployments and deployment status
This application will be able to manage deployments and deployment status (no direct code access).

Personal user data
Email addresses (read-only)
This application will be able to read your private email addresses.

Repository webhooks and services
Read and write access
This application will be able to read and modify repository webhooks and services (no direct code access).
```
which is much more limited.)

I'm still personally a bit hesitant to grant the permissions on this Oauth app at least without a better understanding of what it can do, but there may be good news. The need to install this second application seems to me like a bug. [This page](https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#grant-access-to-your-github-repositories) talks about how the GitHub application should be preferred over the Oauth app for setting up CI. This Oauth app also isn't referenced at all at https://www.azuredevopslabs.com/labs/azuredevops/github-integration/. When you initially create the pipeline following the instructions there, you don't need the Oauth app.

I'm not the only one that thinks this is a bug either, however, it appears Microsoft disagrees as they closed https://developercommunity.visualstudio.com/content/problem/642369/pipelines-creation-falls-back-to-github-oauth-auth.html as "Not a bug". Despite that, they describe a way to work around the problem. You can click the button circled here:

![Screen Shot 2019-09-24 at 4 56 19 PM](https://user-images.githubusercontent.com/6504915/65558627-67679300-deec-11e9-9bb2-457b0512c9b2.png)

This button says it is to create a pipeline without a YAML, but it lets you use your GitHub app connection to link to GitHub and then the first option is to use a YAML. I tested this a bit on my fork and this seems to work.

Do you think these permissions are sufficient for us now and in the future or do you think we will also need/want to set up the Oauth app? I'll flag for you that https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#grant-access-to-your-github-repositories which I linked above says:

> The Azure Pipelines GitHub App is the recommended authentication type for continuous integration pipelines. For release pipelines that are triggered by changes to a GitHub repository, use OAuth or personal access token authentication.

Reply:
Despite the advice in the doc you quoted, I think it is indeed totally doable with this approach. I reconfigured my personal fork, and it is working great in term of CI.

In term of CD, the only question I had was how github releases would be handled in this case. The [documentation](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/github-release?view=azure-devops) states that you need a "Service connection". Following the path for this, it appears that you can typically choose 3 options here:
- use the pre-existing connection that corresponds to the GitHub App installed by Azure. This has been configured at the beginning of connection between Azure Pipelines and GitHub. It has the reasonable set of permissions I wrote at the beginning of the `INSTALL.md` file, section "Installing Azure Pipelines to GitHub".
- create a connection using the "Grant access" option. In this case, you have again an OAuth pop-up permission, like when creating a pipeline in non-classic mode, with the "give me all the rights" thing
- create a connection using the "Personal token access". In this case, recommended (as shown in the UI when trying this option) rights associated to the token to use are: `repo, read:user, user:email, admin:repo_hook`. This is also very reasonable (and really similar to Travis here), with admin control _only_ on the repo hooks. Understandable, since you need to setup several hooks around GitHub releases.

Reply:
Update here: the pre-existing connection cannot be used as a Service connection to push GitHub releases. Azure DevOps complains that it needs to be a Service connection created from a token or OAuth.

So I go for the third option in my previous comment.

Reply:
A personal access token (which should probably be set up using a bot account rather than own of our own GitHub accounts) or just pushing the release outside of Azure (especially if we're going to be signing it locally anyway) seems workable to me.

I'll keep reviewing this PR.

Reply:
Alright. So as discussed, we will continue to work on the continuous delivery using the token approach in the next PR, then properly design the final CD approach we want.